### PR TITLE
All products filter

### PR DESCRIPTION
**Pull request by SStegniy**

### CODE REVIEW

File: figma-shop/src/app/app.module.ts [hunk 1]
```diff
@@ -10,18 +10,20 @@ import { AppRoutingModule } from './app-routing.module';
 import { AppComponent } from './app.component';
 
 import { HeaderModule } from './components/header/header.module';
+import { FilterModule } from './shared/pipes/filter/filter.module';
```
Comment (v-skochko) [inline figma-shop/src/app/app.module.ts:13]: This import produce a building error! There is no directory 'pipes' in this branch

**Review by v-skochko — CHANGES_REQUESTED** [review]
Please fix building errors!

File: figma-shop/src/app/app.module.ts (hunk 1 above)
Comment (SStegniy) [reply figma-shop/src/app/app.module.ts]: Done

**Review by v-skochko — APPROVED** [review]

File: figma-shop/src/app/pages/product-details/details.resolver.ts [hunk 2]
```diff
@@ -2,15 +2,15 @@ import { Injectable } from '@angular/core';
 import { ActivatedRouteSnapshot, Router, Resolve } from '@angular/router';
 import { Observable } from 'rxjs';
 import { ProductInterface } from '../../shared/interfaces/product.interface';
-import { ProductServiceService } from '../../shared/services/product-service.service';
+import { ProductService } from '../../shared/services/product.service';
 
 @Injectable({
   providedIn: 'root'
 })
```
Comment (Vsumskiy) [inline figma-shop/src/app/pages/product-details/details.resolver.ts:9]: remove it, please

File: figma-shop/src/app/pages/products/all-products.resolver.ts [hunk 3]
```diff
@@ -1,14 +1,14 @@
 import { Injectable } from '@angular/core';
 import { ActivatedRouteSnapshot, Resolve } from '@angular/router';
-import { ProductServiceService } from '../../shared/services/product-service.service';
+import { ProductService } from '../../shared/services/product.service';
 import { Observable } from 'rxjs';
 
 @Injectable({
   providedIn: 'root'
 })
```
Comment (Vsumskiy) [inline figma-shop/src/app/pages/products/all-products.resolver.ts:9]: same here

File: figma-shop/src/app/pages/products/filter/filter.component.ts [hunk 4]
```diff
@@ -1,15 +1,115 @@
...
+  private formOnChange(): void {
+    this.form.valueChanges.pipe(distinctUntilChanged()).subscribe(data => {
+      this.filterService.getFilterData({
+        category: data.category,
+        brand: this.checkedBrands,
+        rating: this.checkedRatings,
+        price: [data.price[0], data.price[1]]
+      });
+      this.max = data.price[1];
+      this.min = data.price[0];
```
Comment (Vsumskiy) [inline figma-shop/src/app/pages/products/filter/filter.component.ts:56]: maybe, the better way use destuction?
const [firstElementOfArray, secondElementOfArray] = data.price;

File: figma-shop/src/app/pages/products/filter/filter.component.ts [hunk 5]
```diff
@@ -1,15 +1,115 @@
...
+      rating: this.formBuilder.array(this.allRatings.map(elem => new FormControl(false))),
+      price: new FormControl()
+    });
+  }
+
+  private getCategories(): void {
+    const categories = [];
+    this.allProducts.forEach(prod => categories.push(prod.category.toLowerCase()));
+    this.allCategories = categories.reduce((count, elem) => {
+       count[elem] === undefined ? count[elem] = (count[elem] = 1) : count[elem] = (count[elem] += 1);
```
Comment (Vsumskiy) [inline figma-shop/src/app/pages/products/filter/filter.component.ts:73]: you made here check on 'undefined', if here we get the null?

File: figma-shop/src/app/pages/products/products.component.ts [hunk 6]
```diff
@@ -10,18 +11,36 @@ import { ProductServiceService } from '../../shared/services/product-service.ser
...
-    private productService: ProductServiceService,
-    private actRoute: ActivatedRoute) { }
+    private productService: ProductService,
+    private actRoute: ActivatedRoute,
+    private filterService: FiltersService) { }
 
   ngOnInit(): void {
     this.actRoute.data.subscribe(data => {
       this.allProducts = data.products;
     });
```
Comment (Vsumskiy) [inline figma-shop/src/app/pages/products/products.component.ts:28]: How about unsubscribe?
also, separete it

File: figma-shop/src/app/shared/services/filters.service.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  constructor() { }
+
+  getFilterData(data): void {
+    this.filter = data;
+  }
+
+  filterAllProducts(products: ProductInterface[]): ProductInterface[] {
+    if (!this.filter || this.filter.category === null && this.filter.brand.length === 0 &&
+      this.filter.rating.length === 0) {
```
Comment (Vsumskiy) [inline figma-shop/src/app/shared/services/filters.service.ts:19]: if (!this.filter || this.filter.category === null && this.filter.brand.length &&
      this.filter.rating.length)

File: figma-shop/src/app/shared/services/filters.service.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      return !!product;
+    }
+  }
+
+  private checkPricePresents(product: ProductInterface): boolean {
+    if (this.filter.price?.length) {
+      return this.filter.price[0] <= product.price && this.filter.price[1] >= product.price;
+    }
+    else {
+      return !!product;
```
Comment (Vsumskiy) [inline figma-shop/src/app/shared/services/filters.service.ts:64]: I dont like it '!!'
maybe, the better way Boolean(product)?
But you can leave it as it

File: figma-shop/src/app/pages/products/filter/filter.component.ts (hunk 4 above)
Comment (SStegniy) [reply figma-shop/src/app/pages/products/filter/filter.component.ts:56]: got it

File: figma-shop/src/app/pages/products/filter/filter.component.ts (hunk 5 above)
Comment (SStegniy) [reply figma-shop/src/app/pages/products/filter/filter.component.ts:73]: change on !cout[elem] ? ...

File: figma-shop/src/app/pages/products/products.component.ts (hunk 6 above)
Comment (SStegniy) [reply figma-shop/src/app/pages/products/products.component.ts:28]: OK

File: figma-shop/src/app/shared/services/filters.service.ts (hunk 7 above)
Comment (SStegniy) [reply figma-shop/src/app/shared/services/filters.service.ts:19]: thanks, looks much better

File: figma-shop/src/app/shared/services/filters.service.ts (hunk 8 above)
Comment (SStegniy) [reply figma-shop/src/app/shared/services/filters.service.ts:64]: OK

File: figma-shop/src/app/pages/product-details/details.resolver.ts (hunk 2 above)
Comment (SStegniy) [reply figma-shop/src/app/pages/product-details/details.resolver.ts:9]: Done

File: figma-shop/src/app/pages/products/all-products.resolver.ts (hunk 3 above)
Comment (SStegniy) [reply figma-shop/src/app/pages/products/all-products.resolver.ts:9]: also done

**Review by Vsumskiy — APPROVED** [review]
approved

File: figma-shop/src/app/shared/services/product.service.ts [hunk 9]
```diff
@@ -1,12 +1,13 @@
...
 import { ProductInterface } from '../interfaces/product.interface';
+import { BehaviorSubject } from 'rxjs';
 
 @Injectable({
   providedIn: 'root'
 })
-export class ProductServiceService {
-
+export class ProductService {
+  public products = new BehaviorSubject([]);
```
Comment (maksym-syniuk) [inline figma-shop/src/app/shared/services/product.service.ts:10]:
```suggestion
  public products = new BehaviorSubject<ProductInterface[]>([]);
```

File: figma-shop/src/app/shared/services/product.service.ts [hunk 10]
```diff
@@ -19,6 +20,12 @@ export class ProductServiceService {
   }
 
   public getAllProducts(): Promise<ProductInterface> {
```
Comment (maksym-syniuk) [inline figma-shop/src/app/shared/services/product.service.ts:22]: are you sure that `getAllProducts()` returns `Promise<ProductInterface>`, not `Promise<ProductInterface[]>`?

File: figma-shop/src/app/pages/products/filter/filter.component.ts [hunk 11]
```diff
@@ -1,15 +1,115 @@
...
+      category: new FormControl(),
+      brand: this.formBuilder.array(this.allBrands.map(elem => new FormControl(false))),
+      rating: this.formBuilder.array(this.allRatings.map(elem => new FormControl(false))),
+      price: new FormControl()
+    });
+  }
+
+  private getCategories(): void {
+    const categories = [];
+    this.allProducts.forEach(prod => categories.push(prod.category.toLowerCase()));
```
[the remark this answers is not in the document]
Comment (SStegniy) [reply figma-shop/src/app/pages/products/filter/filter.component.ts:70]: add, even if I pass on an array of ProductInterface[] and IDE knows that it is ProductInterface?

File: figma-shop/src/app/pages/products/filter/filter.component.ts [hunk 12]
```diff
@@ -1,15 +1,115 @@
...
 
-  constructor() { }
+  constructor(
+    private formBuilder: FormBuilder,
+    private filterService: FiltersService,
+    private prodService: ProductService) { }
 
   ngOnInit(): void {
+    this.allProducts = this.prodService.products.getValue();
+    this.getCategories();
```
[the remark this answers is not in the document]
Comment (SStegniy) [reply figma-shop/src/app/pages/products/filter/filter.component.ts:41]: Got it

File: figma-shop/src/app/shared/services/filters.service.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  constructor() { }
+
+  getFilterData(data): void {
+    this.filter = data;
+  }
+
+  filterAllProducts(products: ProductInterface[]): ProductInterface[] {
+    if (!this.filter || this.filter.category === null && this.filter.brand.length === 0 &&
+      this.filter.rating.length === 0) {
+        return products;
```
[the remark this answers is not in the document]
Comment (SStegniy) [reply figma-shop/src/app/shared/services/filters.service.ts:20]: Vasil already tell it, fix it)

File: figma-shop/src/app/shared/services/filters.service.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    } else {
+      return this.checkCategoryPresents(products).filter(product => {
```
[the remark this answers is not in the document]
Comment (SStegniy) [reply figma-shop/src/app/shared/services/filters.service.ts:22]: same question as before

File: figma-shop/src/app/shared/services/product.service.ts (hunk 9 above)
Comment (SStegniy) [reply figma-shop/src/app/shared/services/product.service.ts:10]: ok

File: figma-shop/src/app/shared/services/product.service.ts (hunk 10 above)
Comment (SStegniy) [reply figma-shop/src/app/shared/services/product.service.ts:22]: my mistake, thanks

File: figma-shop/src/app/pages/products/filter/filter.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (SStegniy) [reply figma-shop/src/app/pages/products/filter/filter.component.ts]: Done, thank you

File: figma-shop/src/app/pages/products/filter/filter.component.html [hunk 15]
```diff
@@ -1,122 +1,52 @@
...
-        <input class="filter__checkbox" type="checkbox" name="star">
-        <img class="filter__star" src="./assets/images/star-yell.svg" alt="yellow-star">
-        <img class="filter__star" src="./assets/images/star-yell.svg" alt="yellow-star">
-        <img class="filter__star" src="./assets/images/star-yell.svg" alt="yellow-star">
-        <img class="filter__star" src="./assets/images/star-yell.svg" alt="yellow-star">
-        <img class="filter__star" class="filter__star" src="./assets/images/star-yell.svg" alt="yellow-star">
-      </label>
-    </div>
-    <div class="filter__items">
+    <div class="filter__items" *ngFor="let r of allRatings; let i = index" formArrayName="rating">
```
[the remark this answers is not in the document]
Comment (SStegniy) [reply figma-shop/src/app/pages/products/filter/filter.component.html:21]: Done

File: figma-shop/src/app/pages/products/filter/filter.component.ts [hunk 16]
```diff
@@ -1,15 +1,115 @@
...
   selector: 'app-filter',
   templateUrl: './filter.component.html',
   styleUrls: ['./filter.component.scss']
 })
+
 export class FilterComponent implements OnInit {
+  public objectKeys = Object.keys;
+  private allProducts: ProductInterface[];
+  public allCategories = {};
+  public allBrands = [];
```
[the remark this answers is not in the document]
Comment (SStegniy) [reply figma-shop/src/app/pages/products/filter/filter.component.ts:19]: category is an object which is formed through the reduce method to put the number of products in each category, do i need to create an Interface for him?

**Pull request merged**